python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py

```diff
@@ -993,12 +1012,73 @@ def build_asset_outs(asset_outs: Mapping[str, AssetOut]) -> Mapping[AssetKey, Tu
     return outs_by_asset_key
 
 
-def _make_asset_keys(deps: Optional[Union[Set[AssetKey], Set[str]]]) -> Optional[Set[AssetKey]]:
-    """Convert all str items to AssetKey in the set."""
+def _type_check_deps_and_non_argument_deps(
+    deps: Optional[Sequence[Union[CoercibleToAssetKey, AssetsDefinition, SourceAsset]]],
+    non_argument_deps: Optional[Union[Set[AssetKey], Set[str]]],
+):
+    """Helper function for managing deps and non_argument_deps while non_argument_deps is still an accepted parameter.
+    Ensures:
+    1. only one of deps and non_argument_deps is provided.
+    2. multi assets AssetsDefinition is not passed to deps.
+    3. deprecation warning is fired for non_argument_deps.
+    """
+    if non_argument_deps is not None and deps is not None:
+        raise DagsterInvalidDefinitionError(
+            "Cannot specify both deps and non_argument_deps to @asset. Use only deps instead."
+        )
+
+    upstream_asset_deps: Optional[
+        Sequence[Union[CoercibleToAssetKey, AssetsDefinition, SourceAsset]]
+    ] = None
+    if deps is not None:
+        for dep in deps:
+            # this gnarly type check is because we can't do isinstance(CoercibleToAssetKey) because CoercibleToAssetKey contains a List[str] subtype
```

Comment:
Could we use `AssetKey.from_coercible` here?

Reply:
can you elaborate a bit? use `AssetKey.from_coercible` to do the type check?

Reply:
Yeah or maybe that exact function isn't right. My more general thought here is that, if I understand correctly, we're trying to test whether it will successfully pass `AssetKey.from_coercible`? If that's the case, it seems like this probably isn't the only place that we'll want to be able to check this, and it would be better to put that logic into a method on `AssetKey` than silo it here